codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/RestXmlResponseProtocolGenerator.java

```diff
@@ -0,0 +1,137 @@
+import software.amazon.smithy.typescript.codegen.integration.HttpBindingProtocolGenerator;
+
+/**
+ * Handles general components across the XML-document response based AWS protocols
+ * that have HTTP bindings. It handles reading from document bodies, including
+ * generating any functions needed for performing deserialization.
+ *
+ * This builds on the foundations of the {@link HttpBindingProtocolGenerator} to
+ * handle components of binding to HTTP responses.
+ *
+ * @see XmlShapeDeserVisitor
+ * @see XmlMemberDeserVisitor
+ * @see AwsProtocolUtils
+ * @see <a href="https://awslabs.github.io/smithy/spec/http.html">Smithy HTTP protocol bindings.</a>
+ * @see <a href="https://awslabs.github.io/smithy/spec/xml.html">Smithy XML traits.</a>
+ */
+abstract class RestXmlResponseProtocolGenerator extends HttpBindingProtocolGenerator {
+
+    RestXmlResponseProtocolGenerator() {
+        super(true);
+    }
+
+    @Override
+    protected void generateDocumentBodyShapeDeserializers(GenerationContext context, Set<Shape> shapes) {
+        AwsProtocolUtils.generateDocumentBodyShapeSerde(context, shapes, new XmlShapeDeserVisitor(context));
+    }
+
+    @Override
+    public void generateSharedComponents(GenerationContext context) {
+        super.generateSharedComponents(context);
+
+        TypeScriptWriter writer = context.getWriter();
+
+        // Include an XML body parser used to deserialize documents from HTTP responses.
+        writer.addImport("SerdeContext", "__SerdeContext", "@aws-sdk/types");
+        writer.addDependency(AwsDependency.XML_PARSER);
+        writer.addDependency(AwsDependency.XML_PARSER_TYPES);
+        writer.addImport("parse", "pixlParse", "pixl-xml");
+        writer.openBlock("const parseBody = (streamBody: any, context: __SerdeContext): any => {", "};", () -> {
+            writer.openBlock("return collectBodyString(streamBody, context).then(encoded => {", "});", () -> {
+                writer.openBlock("if (encoded.length) {", "}", () -> {
+                    writer.write("return pixlParse(encoded);");
+                });
+                writer.write("return {};");
+            });
+        });
+
+        writer.write("");
+
+        // Generate a function that handles the complex rules around deserializing
+        // an error code from a rest-xml error.
+        SymbolReference responseType = getApplicationProtocol().getResponseType();
+        writer.openBlock("const loadRestXmlErrorCode = (\n"
+                       + "  output: $T,\n"
+                       + "  data: any\n"
+                       + "): string => {", "};", responseType, () -> {
+            // Start building the location that contains the error code.
+            StringBuilder locationBuilder = new StringBuilder("data.");
+            // Some services, S3 for example, don't wrap the Error object in the response.
+            if (usesWrappedErrorResponse(context)) {
+                locationBuilder.append("Error.");
```

Comment:
Comparing it to V2, where we wrap `Error` with `Errors` also. Maybe it was from query protocol and we mixed them together.

Reply:
I believe that's from the other protocol(s), I'll update this abstraction if necessary later.

Reply:
That's just EC2. Normal query protocols don't have that extra wrapping.